roxie/ccd/ccdqueue.cpp

```diff
@@ -881,6 +894,58 @@ void doPing(IRoxieQueryPacket *packet, const IRoxieContextLogger &logctx)
     output->flush();
 }
 
+//=================================================================================
+
+static ThreadId roxiePacketReaderThread = 0;
+
+class IBYTIbuffer
+{
+    // This class is used to track a finite set of recently-received IBYTI messages, that may have arrived before the messages they refer to
+    // It is accessed ONLY from the main reader thread and as such does not need to be threadsafe (but does need to be fast).
+    // We use a circular buffer, and don't bother removing anything (just treat old items as expired). If the buffer overflows we will end up
+    // discarding the oldest tracked orphaned IBYTI - but that's ok, no worse than if we hadn't tracked them at all.
+    // Probably ends up going funky for a second or so when msTick() wraps, but again, I don't care.
+public:
+    IBYTIbuffer(unsigned _numOrphans) : numOrphans(_numOrphans)
+    {
+        assertex(numOrphans);
+        orphans = new RoxiePacketHeader[numOrphans];
+        tail = 0;
+    }
+    void noteOrphan(const RoxiePacketHeader &hdr)
+    {
+        assert(GetCurrentThreadId()==roxiePacketReaderThread);
+        unsigned now = msTick();
+        // We could trace that the buffer may be too small, if (orphans[tail].activityId >= now)
+        orphans[tail].copy(hdr);
+        orphans[tail].activityId = now + IBYTIbufferLifetime;
+        tail++;
+        if (tail == numOrphans)
+            tail = 0;
+    }
+    bool lookup(const RoxiePacketHeader &hdr) const
+    {
+        assert(GetCurrentThreadId()==roxiePacketReaderThread);
+        unsigned now = msTick();
+        unsigned lookat = tail;
+        do
+        {
+            if (!lookat)
+                lookat = numOrphans;
+            lookat--;
+            if (orphans[lookat].activityId < now)
```

Comment:
marginally better would be

orphans[tail].activityId = now + IBYTIbufferLifetime
and
(now - orphans[lookat].activityId) >=  IBYTIbufferLifetime